src/drivers/imx/sdma.c

```diff
@@ -713,6 +713,9 @@ static int sdma_prep_desc(struct dma_chan_data *channel,
 
 	pdata->next_bd = 0;
 
+	/* Silence "may be uninitialized" warnings at some -O level(s) */
+	bd = &pdata->desc[0]; width = 0;
```

Comment:
This looks ok.
But, can we move the initialization up, where the variables are defined?

Reply:
I don't think this is a good idea but please submit an alternative fix and I'll abandon this one.

I don't think this is a good idea for at least two reasons:
1. I placed these fake initializations immediately before the real one below.
2. I placed these fake initializations _after_ the `config->elem_array.count <= 0` assert that guarantees that `pdata->desc[0]` exists.

As a consequence:
1. makes sure that the compiler turns this into a no-op and optimizes it away, producing the exact same binary as before (except for debug symbols; use `diffoscope` or `-g0` to compare). Having the real and fake initializations close to each other could also make any refactoring easier.
2. makes sure that, if someone ever tries to use `pdata->desc[0]` before the `config->elem_array.count <= 0` assert in the future by mistake, then the compiler can notice and warn about it.

Reply:
My only objection here is to place the initializations on separate lines.

Reply:
even if we put:
```
int width = 0;
struct sdma_bd *bd = NULL;
```
we get the "may be uninitialized" warning?
I believe this is the simplest solution.

Reply:
It's shorter but it's also less safe because it hides _all_ uninitialized warnings, including valid uninitialized warnings. 
So let's pretend someone starts to _actually_ use an uninitialized `bd` after some future refactoring:

```diff
@@ -706,15 +706,18 @@ static int sdma_prep_desc(struct dma_chan_data *channel,
 		return -EINVAL;
 	}
 	if (config->elem_array.count <= 0) {
-		tr_err(&sdma_tr, "sdma_set_config: Invalid descriptor count: %d",
-		       config->elem_array.count);
+		tr_err(&sdma_tr, "sdma_set_config: Invalid descriptor count: %d, %p",
+		       config->elem_array.count, bd);
 		return -EINVAL;
 	}
```

Of course no one would do such a basic mistake, this is just pretending to be some more significant (and more error-prone) refactoring.

With this PR such a mistake is correctly detected by gcc10. With `bd = NULL` at the top then ALL uninitialized uses are silenced, even valid ones like this one.

Of course the even better fix would to BOTH declare and initialization just before the loop, as late as possible just like in any other modern and safer programming language, see updated commit message.